satellite/overlay/nodeselectioncache.go

```diff
@@ -0,0 +1,195 @@
+	Staleness time.Duration `help:"how stale the node selection cache can be" releaseDefault:"3m" devDefault:"5m"`
+}
+
+// NodeSelectionCache keeps a list of all the storage nodes that are qualified to store data
+// We organize the nodes by if they are reputable or a new node on the network.
+// The cache will sync with the nodes table in the database and get refreshed once the staleness time has past.
+type NodeSelectionCache struct {
+	log             *zap.Logger
+	db              CacheDB
+	selectionConfig NodeSelectionConfig
+	staleness       time.Duration
+
+	mu   sync.RWMutex
+	data *state
+}
+
+type state struct {
+	lastRefresh time.Time
+
+	mu             sync.RWMutex
+	reputableNodes []*SelectedNode
+	newNodes       []*SelectedNode
+}
+
+// NewNodeSelectionCache creates a new cache that keeps a list of all the storage nodes that are qualified to store data
+func NewNodeSelectionCache(log *zap.Logger, db CacheDB, staleness time.Duration, config NodeSelectionConfig) *NodeSelectionCache {
+	return &NodeSelectionCache{
+		log:             log,
+		db:              db,
+		staleness:       staleness,
+		selectionConfig: config,
+		data:            &state{},
+	}
+}
+
+// Init populates the cache with all of the reputableNodes and newNode nodes
+// that qualify to upload data from the nodes table in the overlay database
+func (cache *NodeSelectionCache) Init(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	_, err = cache.refresh(ctx)
+	return err
+}
+
+// Refresh calls out to the database and refreshes the cache with the current data
+// from the nodes table, then sets time that the last refresh occurred so we know when
+// to refresh again in the future
+func (cache *NodeSelectionCache) refresh(ctx context.Context) (cachData *state, err error) {
+	defer mon.Task()(&ctx)(&err)
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	if cache.data != nil && time.Since(cache.data.lastRefresh) <= cache.staleness {
+		return cache.data, nil
+	}
+
+	reputableNodes, newNodes, err := cache.db.SelectAllStorageNodesUpload(ctx, cache.selectionConfig)
+	if err != nil {
+		return cache.data, err
+	}
+	cache.data = &state{
+		lastRefresh:    time.Now().UTC(),
+		reputableNodes: reputableNodes,
+		newNodes:       newNodes,
+	}
+
+	mon.IntVal("refresh_cache_size_reputable").Observe(int64(len(reputableNodes)))
+	mon.IntVal("refresh_cache_size_new").Observe(int64(len(newNodes)))
+	return cache.data, nil
+}
+
+// GetNodes selects nodes from the cache that will be used to upload a file.
+// Every node selected will be from a distinct network.
+// If the cache hasn't been refreshed recently it will do so first.
+func (cache *NodeSelectionCache) GetNodes(ctx context.Context, req FindStorageNodesRequest) (_ []*SelectedNode, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	cache.mu.RLock()
+	cacheData := cache.data
+	cache.mu.RUnlock()
+
+	// if the cache is stale, then refresh it before we get nodes
+	if time.Since(cacheData.lastRefresh) > cache.staleness {
```

Comment:
it is a little redundant and we prob don't need this block, we could just check in the refresh method, but i do like the way it reads better. Without this if block, at first glance, it would look like the GetNodes method always refreshes the cache which would be bad

Reply:
No, we definitely need this check. It's this way to avoid using write lock in the common case.

Reply:
oh okay, good to know